### PR TITLE
Added Private IPv4

### PR DESCRIPTION
**Pull request by ZuluPro**
### What does this changes

Added private IPv4 method. I designed it with IP classes (A, B, C)
https://en.wikipedia.org/wiki/Classful_network

### CODE REVIEW

File: faker/providers/internet/__init__.py [hunk 1]
```diff
@@ -178,6 +178,22 @@ def ipv4(self, network=False):
             address = str(ip_network(address, strict=False))
         return address
 
+    def ipv4_private(self, klass=None, network=False):
```
[the remark this answers is not in the document]
Comment (fcurella) [reply faker/providers/internet/__init__.py]: Could we merge this logic into the `ipv4` method and have non-private address classes? I'm thinking something like:

```python
def ipv4(self, network=False, private=None, address_class=None):
    if private is None and address_class is None:
        # pretty much the old logic
    if private is True:
        classes = {
            'a': (167772160, 184549375, 24),
            'b': (2886729728, 2887778303, 20),
            'c': (3232235520, 3232301055, 16),
        }
  if private is False:
        classes = {
            'b': (2886729728, 2887778303, 16),
            'c': (3232235520, 3232301055, 8),
        }
```

[the remark this answers is not in the document]
Comment (fcurella) [reply faker/providers/internet/__init__.py]: should we have the same for IPV6 addresses?

**Comment by fcurella** [comment]
alternative implementation of https://github.com/joke2k/faker/pull/703

File: faker/providers/internet/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ZuluPro) [reply faker/providers/internet/__init__.py]: > I would use address_class instead of klass
Ok

> Could we merge this logic into the ipv4 method and have non-private address classes?
I don't think we should merge Unaware/Public/Private. 

> should we have the same for IPV6 addresses?
Yes of course, but another PR ?

[the remark this answers is not in the document]
Comment (fcurella) [reply faker/providers/internet/__init__.py]: > I don't think we should merge Unaware/Public/Private.

Could you elaborate why that would be a bad idea? I'd like to keep the number of methods to a minimum

[the remark this answers is not in the document]
Comment (ZuluPro) [reply faker/providers/internet/__init__.py]: Yes, of course,
1. To get private or public is a little complex and separated code is better.
2. Like the `person` provider with male & female, we should provide simple way to get private or public IP.

**Comment by ZuluPro** [comment]
I'm doing `ipv4_public`

**Comment by ZuluPro** [comment]
@fcurella 
What do you think about the code ?
Do you know how to fix this Python 2 bug ?

File: faker/providers/internet/__init__.py [hunk 2]
```diff
@@ -178,6 +210,45 @@ def ipv4(self, network=False):
...
+            address = str(ip_network(address, strict=False))
+        return address
+
+    def ipv4_public(self, network=False, address_class=None):
+        address_class = address_class or self.ipv4_network_class()
+        public_net = IPV4_PUBLIC_NETS[address_class]
+        private_blocks = IPV4_PRIVATE_NET_BLOCKS[address_class]
+        # Make valid IP ranges
+        net_ranges = []
+        if public_net[0] != private_blocks[0][0]:
```
Comment (fcurella) [inline faker/providers/internet/__init__.py]: Could you explain why do we need this check? This code is quite hard to follow, comments would help

File: faker/providers/internet/__init__.py [hunk 3]
```diff
@@ -178,6 +210,45 @@ def ipv4(self, network=False):
             address = str(ip_network(address, strict=False))
         return address
 
+    def ipv4_network_class(self):
+        return self.random_element('abc')
+
+    def ipv4_private(self, address_class=None, network=False):
```
Comment (fcurella) [inline faker/providers/internet/__init__.py]: eventually I want to remove `ipv4_private` and `ipv4_public` from the api in favor of a `private=None` argument for `ipv4`

**Comment by fcurella** [comment]
> Do you know how to fix this Python 2 bug

I think `py2-ipaddress` is our best option in terms of maintainability.

**Comment by fcurella** [comment]
On a second thought, you could use `force_text` from `faker.utils.text`

File: faker/providers/internet/__init__.py (hunk 3 above)
Comment (fcurella) [reply faker/providers/internet/__init__.py]: n/m this comment, I've read a little bit and know I understand that private networks are in multiple classes

Comment (fcurella) [reply faker/providers/internet/__init__.py]: I think the best way to go about this is to have:
```python
def ipv4(self, network=False, address_class=None, private=None):
    if address_class is not None or private is not None:
        if private is True:
            return self.ipv4_private(address_class=address_class, network=network)
        else:
            return self.ipv4_public(address_class=address_class, network=network)
    # old logic here...
```

File: faker/providers/internet/__init__.py (hunk 2 above)
Comment (ZuluPro) [reply faker/providers/internet/__init__.py]: I check if the 1st available range starts after the 1st private block.
Example:
Class C: 192.0.0.0/24
The first address 192.0.0.0 is in 192.0.0.0/8 which is private.
So the 1st `net_ranges` will be from 192.0.0.8 to 192.X.X.X

I'll add comments

**Comment by ZuluPro** [comment]
Ok,
- Comments added
- Docstring added
- Test added
- `ipv4` improved
- Python 2 fixed with `text.force_text`

File: faker/providers/internet/__init__.py [hunk 4]
```diff
@@ -169,15 +201,90 @@ def url(self):
...
+        :param private: Public or private
+        :returns: IPv4
+        """
+        if private is True:
+            return self.ipv4_private(address_class=address_class,
+                                     network=network)
+        elif private is False:
+            return self.ipv4_public(address_class=address_class,
+                                    network=network)
         address = str(ip_address(self.generator.random.randint(
```
Comment (fcurella) [inline faker/providers/internet/__init__.py]: can we add the option to return any IP (without checking if public or private) from a specific address_class?

Comment (ZuluPro) [reply faker/providers/internet/__init__.py]: I prefer to not provider address class for `ipv4()`, because I would like this method to provide any IP.
From private/public, from reserved blocks, etc

Comment (fcurella) [reply faker/providers/internet/__init__.py]: the `address_class` will optional, and default to the old algorithm of returning any IP. We would filter by address class only if `address_class` is specified

**Comment by ZuluPro** [comment]
@fcurella  Changed !

**Comment by fcurella** [comment]
Thank you so much for working on this! ✨
